File: May-LeetCoding-Challange-2020/Week1/278-first-bad-version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution:
    # Sol 1. Linear Search
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        for i in range(n):
            if(isBadVersion(i)):
                return i
        return n
        # Time Complexity: O(n)
        # Space Complexity: O(1)
        # Working, Not accepted, time-limit exceeded

    # Sol 2. Binary Search Tree
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        left = 1 
        right = n
        
        while(left<right):
            mid = left+(right-left)//2 # In Python 3, / returns floating-point and // returns integer 
            
            if not isBadVersion(mid):
                left = mid + 1
            else: 
                right = mid
        
        return left
        # Time Complexity: O(log(n))
        # Space Complexity: O(1)

File: May-LeetCoding-Challange-2020/Week1/278-first-bad-version.cpp
class Solution {
public:
    int firstBadVersion(int n) {
        
        /*
         * Time Complexity: O(logn)
         * Space Complexity: O(1)
         * Median Formula: start + (end-start)/2
         * 
         */
        int start = 0;
        int end = n;
        int pivot = n;

        if(n == 1)
            return n;
        
        while(1){
            
            if(isBadVersion(pivot)){
                if(!isBadVersion(pivot-1))
                    return pivot;
                
                end = pivot;
                pivot = start + (end - start)/2;
                
            }else{
    
                start = pivot;
                pivot = start + (end - start)/2;
            }
        }
    }
};

class Solution {
public:
    int firstBadVersion(int n) {
        
        /*
         * 
         * TIME LIMIT EXCEEDED
         * Reason: Wrong formula for getting median 
         * 
         */
        
        int start = 0;
        int end = n;
        int pivot = n;
        
        if(n == 1)
            return n;
        
        while(1){
            
            if(isBadVersion(pivot)){
                if(!isBadVersion(pivot-1))
                    return pivot;
                
                pivot -= (pivot-start)/2;
                
            }else{
            
                pivot += (end-pivot)/2;
            }
        }
            
    }
};